Simplify chalk-engine a bit

jackh726 opened the pull request:
cc https://github.com/rust-lang/chalk/issues/255

This removes the two different `Context`s from `Strand` and the `&'table mut dyn InferenceTable<C, I>` field (and consequently the lifetime parameter `'table`) has been removed, in lieu of passing `infer` directly to function arguments as needed. `WithInstantiatedUCanonicalGoal` and `WithInstantiatedExClause` were also removed in favor of just `FnOnce`s. And then the `with_instantiated_strand` indirection has also been removed.

`CanonicalStrand`s are still stored in `Table`, since I haven't yet figured out the best way to simplify that. But this gets us one step closer.

I'll probably still be playing around to try to remove `CanonicalStrand`, but I figured it would be better to at least make this PR since it seems fairly straightforward.

nikomatsakis commented:
Yep, it looks like this is the commit to look at

5d8bc3360a7bd4b2ae14f2b386ca744cd859ab0c

jackh726 commented:
@nikomatsakis Yes, I went through the history and found that. The problem is now we _can't_ store the `InferenceTable` in the Strand, since it now has a lifetime associated with it (from `&'me dyn RustIrDatabase`).

nikomatsakis commented:
@jackh726 I see. Interesting! Well, I think we can land this PR for now and look into this question as follow-up. The only thing I'm not sure about is whether it makes sense to remove the inference table from `Strand`, as I think we might likely want to put it back again.

nikomatsakis commented:
I suspect that the right answer here is going to be:

* move `program_clauses` from the `InferenceTable` trait to `ContextOps`
* then we can remove the field `program` from the `TruncatingInferenceTable`
* then we can store it in `Strand`

jackh726 commented:
@nikomatsakis I can definitely try that later tonight. I can also start topic on Zulip if you want to discuss more in sync.

nikomatsakis commented:
@jackh726 zulip topic sounds good =)

jackh726 commented on chalk-engine/src/context.rs:
```diff
@@ -401,3 +374,9 @@ pub trait AnswerStream<C: Context> {
         -> bool;
 }
+
+pub trait Canonical<T, C: Context> {
```
I was looking at some of the git history since writing this, and it seems like there was a push to reduce the traits in `Context`, so I'm not sure this trait makes since.

nikomatsakis replied:
Yeah, not sure, I did push away from these sorts of traits at some point. I think it was because, for real flexibility, you wind up wanting to thread `&C` arguments around anyway, since some crates may need access to fields.

nikomatsakis replied:
Oh and I think it was because of coherence reasons, too -- this way, you only needed to own the "context" type, but the other types might be from other crates

nikomatsakis replied:
I think I'd prefer to have the methods collected on the `Context` type, but we could introduce "indirection helpers" if desired to make the code read more nicely. Still, I'm not convinced these methods are needed at all.

nikomatsakis commented on chalk-engine/src/context.rs:
```diff
@@ -258,20 +239,20 @@ pub trait InferenceTable<C: Context, I: Context>:
 
     /// Upcast this domain goal into a more general goal.
-    fn into_goal(&self, domain_goal: I::DomainGoal) -> I::Goal;
+    fn into_goal(&self, domain_goal: C::DomainGoal) -> C::Goal;
```
An aside: I think it may be worth creating a first PR that **just** removes the `I: Context` type parameters and collapses to one `C: Context` everywhere, but leaves everything else the same. It'd be good to pull out and land intermediate steps.

nikomatsakis replied:
Well, we could probably get rid of the `WithFoo` traits and convert them to `impl FnOnce`, too, but leave everything else about them the same.

nikomatsakis replied:
I guess this was fairly close to your original PR :)

nikomatsakis commented on chalk-engine/src/logic.rs:
```diff
@@ -440,22 +367,9 @@ impl<C: Context> Forest<C> {
-            *canonical_strand = delayed_strand;
-
+        for strand in self.tables[table].strands_mut() {
+            // FIXME: do we need to uncanonicalize this
```
I don't think so -- but I have to review this logic -- in any case I actually want to remove this "delaying" stuff, see #250, so maybe that too would be a good step to take before doing the rest of this PR, just to simplify things?

jackh726 commented:
@nikomatsakis Ok, so I removed the "remove `CanonicalStrand`" commit, and added a couple other smaller cleanup changes. I do think that any followup work for removing `CanonicalStrand` should be in another PR, possibly after removing the delaying stuff.

Let me know if you disagree with the other smaller changes I made and I can just remove them.
